lib/aws/request.js

```diff
@@ -0,0 +1,234 @@
+            ) {
+              const nextTryNum = numTry + 1;
+              const jitter = Math.random() * 3000 - 1000;
+              // backoff is between 4 and 7 seconds
+              const backOff = BASE_BACKOFF + jitter;
+
+              log(
+                [
+                  `Recoverable error occurred (${e.message}), sleeping for ~${Math.round(
+                    backOff / 1000
+                  )} seconds.`,
+                  `Try ${nextTryNum} of ${MAX_RETRIES}`,
+                ].join(' ')
+              );
+              setTimeout(doCall, backOff, nextTryNum);
+            } else {
+              reject(e);
+            }
+          });
+      };
+      return doCall(0);
+    });
+
+  const request = requestQueue.add(() =>
+    persistentRequest(() => {
+      const awsService = getServiceInstance(service, method);
+      const req = awsService[method](...args);
+
+      // TODO: Add listeners, put Debug statements here...
+      // req.on('send', function (r) {console.log(r)});
+      const promise = req.promise();
+      return promise.catch((err) => {
+        let message = err.message != null ? err.message : String(err.code);
+        if (message.startsWith('Missing credentials in config')) {
+          // Credentials error
+          // If failed at last resort (EC2 Metadata check) expose a meaningful error
+          // with link to AWS documentation
+          // Otherwise, it's likely that user relied on some AWS creds, which appeared not correct
+          // therefore expose an AWS message directly
+          let bottomError = err;
+          while (bottomError.originalError && !bottomError.message.startsWith('EC2 Metadata')) {
+            bottomError = bottomError.originalError;
+          }
+
+          const errorMessage = bottomError.message.startsWith('EC2 Metadata')
+            ? [
+                'AWS provider credentials not found.',
+                ' Learn how to set up AWS provider credentials',
+                ` in our docs here: <${chalk.green('http://slss.io/aws-creds-setup')}>.`,
+              ].join('')
+            : bottomError.message;
+          message = errorMessage;
+          // We do not want to trigger the retry mechanism for credential errors
+          return Promise.reject(
+            Object.assign(new ServerlessError(errorMessage, err.code), {
+              providerError: Object.assign({}, err, { retryable: false }),
+            })
+          );
+        }
+
+        return Promise.reject(
+          Object.assign(new ServerlessError(message, err.code), {
+            providerError: err,
+          })
+        );
+      });
+    }).then((data) => {
+      const result = Promise.resolve(data);
+      return result;
+    })
+  );
+
+  return request;
+}
+
+awsRequest.memoized = memoize(awsRequest, {
+  normalizer: ([service, method, args]) => {
+    return [deepSortObjectByKey(service), method, deepSortObjectByKey(args)].join('|');
```

Comment:
We need to stringify arguments (Possibly best would be to `[].map(value => JSON.stringify(value)).join("|")`). Also I think `service` can be a string (?)

Reply:
At that point, not anymore, I've moved the type assertions to make it clearer.

Reply:
After checking a bit, this is also the case for the args, so I've modified the implementation to include the stringify.